griptape/tools/google_drive/tool.py

```diff
@@ -0,0 +1,293 @@
+import logging
+from schema import Schema, Literal, Optional
+from attr import define, field
+from griptape.artifacts import ErrorArtifact, InfoArtifact, ListArtifact, BlobArtifact
+from griptape.utils.decorators import activity
+from griptape.tools import BaseGoogleClient, BaseTool
+from google.auth.exceptions import MalformedError
+from googleapiclient.discovery import Resource
+from googleapiclient.http import MediaIoBaseUpload
+from googleapiclient.errors import HttpError
+from io import BytesIO
+
+@define
+class GoogleDriveClient(BaseGoogleClient, BaseTool):
+    LIST_FILES_SCOPES = ['https://www.googleapis.com/auth/drive.readonly']
+
+    UPLOAD_FILE_SCOPES = ['https://www.googleapis.com/auth/drive.file']
+
+    GOOGLE_EXPORT_MIME_MAPPING = {
+        "application/vnd.google-apps.document":
+            "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
+        "application/vnd.google-apps.spreadsheet":
+            "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
+        "application/vnd.google-apps.presentation":
+            "application/vnd.openxmlformats-officedocument.presentationml.presentation"
+    }
+
+    owner_email: str = field(kw_only=True)
+    page_size: int = field(default=100, kw_only=True)
+
+    def _build_client(self, scopes: list[str]) -> Resource:
+        from google.oauth2 import service_account
+        from googleapiclient.discovery import build
+
+        credentials = service_account.Credentials.from_service_account_info(
+            self.service_account_credentials, scopes=scopes
+        )
+        delegated_credentials = credentials.with_subject(self.owner_email)
+        service = build('drive', 'v3', credentials=delegated_credentials)
+
+        return service
+
+    def _path_to_file_id(self, service, path: str) -> Optional[str]:
+        parts = path.split('/')
+        current_id = 'root'
+
+        for index, part in enumerate(parts):
+            if index == len(parts) - 1:
+                query = f"name='{part}' and '{current_id}' in parents"
+            else:
+                query = f"name='{part}' and '{current_id}' in parents and mimeType='application/vnd.google-apps.folder'"
+
+            response = service.files().list(q=query).execute()
+            files = response.get('files', [])
+
+            if not files:
+                return None
+            current_id = files[0]['id']
+
+        return current_id
+
+    @activity(config={
+        "description": "List files in a specific Google Drive folder or the root if no folder is specified",
+        "schema": Schema({
+            Optional("folder_path",
+                     description="Path of the Google Drive folder (like 'MainFolder/Subfolder1/Subfolder2') "
+                                 "from which files should be listed. If not provided, files from the root "
+                                 "directory will be listed."): str,
+            Optional("max_files",
```

Comment:
So, answering your previous question, I think this shouldn't be up to the LLM. The goal of the activity is to show all files in the directory, not a subset of files.

We should return all files in the directory, so it's unambiguous. That means that the method that pulls files needs to have pagination logic built-in. In that case, we don't even need to expose `page_size` as a property because it will be just an implementation details specific to your implementation.

Reply:
Done!